lib/SILGen/SILGenApply.cpp

```diff
@@ -1595,6 +1595,39 @@ class SILGenApply : public Lowering::ExprVisitor<SILGenApply> {
     return std::move(*applyCallee);
   }
 
+  /// \returns true if the conversion is from an async function to
+  /// the same type but with a global actor added. For example this:
+  ///     () async -> ()  ==>  @MainActor () async -> ()
+  /// will return true. In all other cases, returns false.
+  static bool addsGlobalActorToAsyncFn(FunctionConversionExpr *fce) {
```

Comment:
I wonder if the same issue can crop up when referencing a `@preconcurrency` function with `@Sendable` parameters from outside of a concurrency context, because we also put in function conversions to cover that case. For example:

```swift
@objc protocol P {
  @preconcurrency @objc optional func f(_: Sendable)
}

func test(p: any P, a: Any) {
  p.f?(a)
  p.f!(a)
 // ...
}
```

Reply:
We have a more general issue with optional-calls to async optional ObjC protocol requirements that I'm tracking in rdar://98465968